internal/graph/singleflight_resolver_test.go

```diff
@@ -0,0 +1,95 @@
+	"github.com/openfga/openfga/pkg/storage/memory"
+	"github.com/openfga/openfga/pkg/storage/storagewrappers"
+	"github.com/openfga/openfga/pkg/tuple"
+	"github.com/openfga/openfga/pkg/typesystem"
+	"github.com/stretchr/testify/require"
+)
+
+func TestSingleflightResolver(t *testing.T) {
+	ds := memory.New()
+	defer ds.Close()
+
+	storeID := ulid.Make().String()
+
+	var tuples = []*openfgav1.TupleKey{
+		tuple.NewTupleKey("folder:1", "viewer", "user:jon"),
+	}
+	for i := 1; i <= 5; i++ {
+		tuples = append(tuples, tuple.NewTupleKey(fmt.Sprintf("folder:%d", i+1), "parent", fmt.Sprintf("folder:%d", i)))
+		tuples = append(tuples, tuple.NewTupleKey(fmt.Sprintf("folder:%d", i+1), "other_parent", fmt.Sprintf("folder:%d", i)))
+		tuples = append(tuples, tuple.NewTupleKey(fmt.Sprintf("folder:%d", i+1), "other_other_parent", fmt.Sprintf("folder:%d", i)))
+	}
+
+	err := ds.Write(context.Background(), storeID, nil, tuples)
+	require.NoError(t, err)
+
+	typedefs := parser.MustTransformDSLToProto(`model
+	schema 1.1
+	
+	type user
+	
+	type folder
+	relations
+	  define parent: [folder]
+	  define other_parent: [folder]
+	  define viewer: [user] or viewer from parent or viewer from other_parent or viewer from other_other_parent`).TypeDefinitions
+
+	ctx := typesystem.ContextWithTypesystem(context.Background(), typesystem.New(
+		&openfgav1.AuthorizationModel{
+			Id:              ulid.Make().String(),
+			TypeDefinitions: typedefs,
+			SchemaVersion:   typesystem.SchemaVersion1_1,
+		},
+	))
+
+	checkReq := ResolveCheckRequest{
+		StoreID:            storeID,
+		TupleKey:           tuple.NewTupleKey("folder:5", "viewer", "user:jon"),
+		ContextualTuples:   nil,
+		ResolutionMetadata: &ResolutionMetadata{Depth: 5},
+	}
+
+	//--------------------------------------
+
+	checkerWithoutSingleflight := NewLocalChecker(
+		storagewrappers.NewCombinedTupleReader(ds, []*openfgav1.TupleKey{}),
+	)
+	defer checkerWithoutSingleflight.Close()
+
+	req := checkReq
+	respWithoutSingleflight, err := checkerWithoutSingleflight.ResolveCheck(ctx, &req)
+
+	require.NoError(t, err)
+	require.False(t, respWithoutSingleflight.GetResolutionMetadata().TMP_Singleflight.HadSharedRequest)
+	require.True(t, respWithoutSingleflight.GetAllowed())
+
+	fmt.Println("Without singleflight:")
+	fmt.Println(fmt.Sprintf("%+v", respWithoutSingleflight.GetResolutionMetadata()))
+
+	//--------------------------------------
+
+	checkerWithSingleflight := NewLocalChecker(
+		storagewrappers.NewCombinedTupleReader(ds, []*openfgav1.TupleKey{}),
+		WithSingleflightResolver(),
+	)
+	defer checkerWithSingleflight.Close()
+
+	req = checkReq
+	resWithSingleflight, err := checkerWithSingleflight.ResolveCheck(ctx, &req)
+
+	fmt.Println("With singleflight:")
+	fmt.Println(fmt.Sprintf("%+v", resWithSingleflight.GetResolutionMetadata()))
+	require.NoError(t, err)
+	require.True(t, resWithSingleflight.GetResolutionMetadata().TMP_Singleflight.HadSharedRequest)
+	require.True(t, resWithSingleflight.GetAllowed())
+}
```

Comment:
These tests pass, however when viewing the log output, it is clear that the singleflight resolver actually _increases_ the number of DB calls when compared to the normal check resolver. This is counter-productive to the intent of this PR and still needs to be investigated why.

```
=== RUN   TestSingleflightResolver
Without singleflight:
&{Depth:0 DatastoreQueryCount:16 TMP_Singleflight:{NumInnerCalls:0 NumOuterCalls:0 HadSharedRequest:false}}
With singleflight:
&{Depth:0 DatastoreQueryCount:14 TMP_Singleflight:{NumInnerCalls:1 NumOuterCalls:9 HadSharedRequest:true}}
--- PASS: TestSingleflightResolver (0.00s)
```

Reply:
This behavior that you are seeing is normal, but I do understand that it is confusing at best. Here are a couple of things to remember:

1. Any de-duplicated `ResolveCheck` calls are not simply discarded, but they use the same result from the first invocation of that overlapping subproblem. The devil is in the details of the godoc for `singleflight.Group#Do`, namely

>  If a duplicate comes in, the duplicate caller waits for the original to complete and receives the same results.

So the datastore query counter will still accumulate and sum up the received results, because each de-duplicated query gets the same result for the one that was resolved. But, that doesn't mean we _actually_ called the storage layer. So the resolution metadata is a bit misleading in that regard. Technically speaking the actually datastore query count is equal to the number of queries / (# de-deduplicated requests), but the metrics don't account for that because of how the composition works.

Put differently, we're not discarding the functions being called, we're just re-using a single shared response. The distinction is nuanced but very important.

2. To accurately account for the reduction of compute that `singleflight` introduces, we'll need to actually count the number of concurrent/overlapping invocations of `singleflight.Group#Do` on the same key (not just a shared boolean). See https://github.com/openfga/openfga/pull/1313/files#r1462591562 for more information.

Reply:
Great explanation. Thanks!